Add S3-URI as upload path style

sbkok opened the pull request:
In order to reference files from CloudFormation, like a Swagger
OpenAPI definition, you need to use the `s3://` format.

At the moment, the `upload:` helper is able to return `path` or
`virtual-hosted` urls. With this PR, it will get support to return
S3-URIs as well.

Modifications:

- Added S3-URI syntax support.
- Removed redundant code for `path` and `virtual-hosted`.
- Added tests to verify the styles and different region configs.
- Documentation has been updated to clearly distinguish the different
  output styles.

**References:**

- Fixes #197 (requests `s3://` as an output format for `upload:` helper).

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

sbkok commented:
@bundyfx While writing the tests, I noticed that `virtual-hosted` style is using `http://` as its output protocol, not `httpS://`.

Is this for any particular reason that I am not aware of, or should I change that to become `https://` as well?

bundyfx commented:
> @bundyfx While writing the tests, I noticed that `virtual-hosted` style is using `http://` as its output protocol, not `httpS://`.
> 
> Is this for any particular reason that I am not aware of, or should I change that to become `https://` as well?

@sbkok Should be using https:// for sure! Good spotting, if you want to make that change this looks good to go, awesome work and thank you very much!

sbkok commented:
@bundyfx Thanks for your quick response. I've updated the PR.

Today, @thomasmcgannon shared with me that CodeBuild project source locations refer to S3 objects without the `s3://` protocol. So  I've split the setup into two styles:

- `s3-uri` which returns: `${bucket}/${key}`.
- `s3-url` which returns: `s3://${bucket}/${key}`.
